lib/viewmodels/create_accout_view_model.dart

```diff
@@ -0,0 +1,120 @@
+  /// - It is more than 20 characters long
+  /// - It contains special characters
+  String? _validateString(String value) {
+    if (value.isEmpty) {
+      return "Cannot be blank.";
+    }
+
+    if (value.contains(" ")) {
+      return "Cannot contain spaces.";
+    }
+
+    if (value.length < 3) {
+      return "Too short.";
+    }
+
+    if (value.length > 20) {
+      return "Too long.";
+    }
+
+    // This regular expression is intentionally too restrictive (for instance,
+    // the {3,20} is already checked above). Its purpose is too make sure that
+    // the value is whitelisted; not that it is not blacklisted.
+    if (!RegExp(r"^\w{3,20}$").hasMatch(value)) {
+      return "Invalid characters.";
+    }
+
+    return null;
+  }
+
+  /// Validate a [pseudo] by returning an error message if it is invalid. Returns null if it is valid.
+  /// A pseudo is invalid if:
+  /// - It is invalid according to [_validateString]
+  String? validatePseudo(String pseudo) {
+    return _validateString(pseudo);
+  }
+
+  /// Validate a [uniqueUsername] by returning an error message if it is invalid. Returns null if it is valid.
+  /// A unique username is invalid if:
+  /// - It is invalid according to [_validateString]
+  /// - It is already taken by another user in the database
+  Future<String?> validateUniqueUsername(String uniqueUsername) async {
+    final error = _validateString(uniqueUsername);
+    if (error != null) {
+      return error;
+    }
+
+    if (await ref
+        .read(userRepositoryProvider)
+        .isUsernameTaken(uniqueUsername)) {
+      return "This username is already taken.";
+    }
+    return null;
+  }
+
+  /// Validate a [pseudo] and a [uniqueUsername] and update the state with the potential errors.
+  /// If it happens that both are valid, the account is created.
+  /// See [validatePseudo] and [validateUniqueUsername] for the validation rules.
+  Future<void> createAccountIfValid(
+    String pseudo,
+    String uniqueUsername,
+  ) async {
+    state = const AsyncValue.loading();
+    final newState = await AsyncValue.guard(() async {
+      final pseudoError = validatePseudo(pseudo);
+      final uniqueUsernameError = await validateUniqueUsername(uniqueUsername);
+      return CreateAccountErrors(
+        pseudoError: pseudoError,
+        uniqueUsernameError: uniqueUsernameError,
+        valid: pseudoError == null && uniqueUsernameError == null,
+      );
+    });
+
+    // Create the account before applying the new state
+    if (newState.valueOrNull?.valid == true) {
+      final uid = ref.read(uidProvider);
+      if (uid == null) {
+        // The user is no longer logged in, so they will anyway be sent to the login page
+        state = const AsyncValue.data(null);
+        return;
+      }
+
+      final userData = UserData(
+        username: uniqueUsername,
+        displayName: pseudo,
+        joinTime: Timestamp.now(),
+      );
+      ref.read(userRepositoryProvider).setUser(uid, userData);
```

Comment:
I believe we should add an `await` to make sure that the function returns only after the user has been added.
```suggestion
      await ref.read(userRepositoryProvider).setUser(uid, userData);
```